File: runtime/frame_codec.py
"""
Pure frame codec + reassembler for the ``tywrap-frame/1`` framing protocol.

This is the Python mirror of ``src/runtime/frame-codec.ts``. It fragments one
complete logical JSON message into chunk frames and reassembles a stream of
frames back into the original string. It performs NO I/O and reads NO env vars
(PURITY, matching ``tywrap_bridge_core``): the read/write loop in
``python_bridge.py`` (W4/W5) wires these functions onto stdin/stdout.

Encoding is ``utf8-slice`` (plan decision #6, docs/transport-framing.md): the
logical payload is already valid-UTF-8 JSON, so each frame's ``data`` is a raw
substring split on a UTF-8 codepoint boundary at or before ``max_frame_bytes``
UTF-8 bytes. Reassembly is plain concatenation -- no base64, no ~33% inflation.
A frame's ``data`` MUST NOT split a multi-byte UTF-8 sequence; ``encode_frames``
guarantees this by snapping every boundary back to the nearest codepoint
boundary.

The two implementations MUST agree byte-for-byte on the wire (see
test/python/test_frame_codec.py and test/frame-codec.test.ts for the
cross-language parity vectors).
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

# Framing protocol identifier; mirrors FRAME_PROTOCOL_ID in
# src/runtime/transport.ts. Any other value on the wire is rejected.
FRAME_PROTOCOL_ID = 'tywrap-frame/1'

# Numeric framing-protocol version, derived from the trailing number so the two
# cannot drift (same pattern as PROTOCOL_VERSION in tywrap_bridge_core).
FRAME_PROTOCOL_VERSION = int(FRAME_PROTOCOL_ID.split('/')[1])

# The single per-frame encoding tywrap emits/accepts in 0.8.0.
FRAME_ENCODING = 'utf8-slice'


class FrameError(Exception):
    """A framing-protocol violation (malformed/duplicate/inconsistent frame).

    Carries a stable ``code`` mirroring the BridgeProtocolError codes on the TS
    side so callers can branch on the failure class.
    """

    def __init__(self, message: str, *, code: str) -> None:
        super().__init__(message)
        self.code = code


def _utf8_bytes_of_codepoint(code_point: int) -> int:
    """UTF-8 byte length of a single Unicode codepoint."""
    if code_point <= 0x7F:
        return 1
    if code_point <= 0x7FF:
        return 2
    if code_point <= 0xFFFF:
        return 3
    return 4


def utf8_byte_length(value: str) -> int:
    """Exact UTF-8 byte length of a string."""
    return len(value.encode('utf-8'))


def encode_frames(
    logical_json: str,
    *,
    id: int,
    stream: str,
    max_frame_bytes: int,
) -> List[Dict[str, Any]]:
    """Fragment a complete logical JSON message into ``tywrap-frame/1`` frames.

    Splits ``logical_json`` on UTF-8 codepoint boundaries so each frame's
    ``data`` is at most ``max_frame_bytes`` UTF-8 bytes and never splits a
    multi-byte sequence. ``totalBytes`` is the exact UTF-8 byte length of the
    full message; ``total`` is the resulting frame count; ``seq`` is zero-based
    and dense.

    An empty ``logical_json`` still produces exactly one (empty) frame so the
    receiver always sees ``total >= 1`` and a well-formed stream.

    :raises FrameError: if ``max_frame_bytes`` is not an int >= 4, or ``stream``
        is not ``"request"``/``"response"``.
    """
    if not isinstance(max_frame_bytes, int) or isinstance(max_frame_bytes, bool) or max_frame_bytes < 4:
        raise FrameError(
            f'encode_frames: max_frame_bytes must be an int >= 4 (got {max_frame_bytes!r})',
            code='FRAME_BAD_MAX_BYTES',
        )
    if stream not in ('request', 'response'):
        raise FrameError(
            f'encode_frames: stream must be "request" or "response" (got {stream!r})',
            code='FRAME_MALFORMED',
        )

    total_bytes = utf8_byte_length(logical_json)

    # Walk codepoints, accumulating UTF-8 bytes into the current slice until the
    # next codepoint would exceed max_frame_bytes; that boundary is, by
    # construction, a codepoint boundary so no multi-byte sequence is split.
    slices: List[str] = []
    current_chars: List[str] = []
    current_bytes = 0
    for ch in logical_json:
        ch_bytes = _utf8_bytes_of_codepoint(ord(ch))
        if current_bytes + ch_bytes > max_frame_bytes and current_chars:
            slices.append(''.join(current_chars))
            current_chars = []
            current_bytes = 0
        current_chars.append(ch)
        current_bytes += ch_bytes
    # Always emit a final slice (covers the empty-string case: one empty frame).
    slices.append(''.join(current_chars))

    total = len(slices)
    return [
        {
            '__tywrap_frame__': 'chunk',
            'frameProtocol': FRAME_PROTOCOL_ID,
            'stream': stream,
            'id': id,
            'seq': seq,
            'total': total,
            'totalBytes': total_bytes,
            'encoding': FRAME_ENCODING,
            'data': data,
        }
        for seq, data in enumerate(slices)
    ]


def parse_chunk_frame(value: Any) -> Dict[str, Any]:
    """Validate that ``value`` is a structurally well-formed data frame.

    Returns the frame dict, or raises :class:`FrameError`. Only
    ``__tywrap_frame__ == "chunk"`` frames flow through reassembly; ``"error"``
    frames are a transport-layer concern handled above this module.
    """
    if not isinstance(value, dict):
        raise FrameError('frame: expected an object', code='FRAME_MALFORMED')

    if value.get('__tywrap_frame__') != 'chunk':
        raise FrameError(
            f'frame: __tywrap_frame__ must be "chunk" (got {value.get("__tywrap_frame__")!r})',
            code='FRAME_MALFORMED',
        )
    if value.get('frameProtocol') != FRAME_PROTOCOL_ID:
        raise FrameError(
            f'frame: unknown frameProtocol {value.get("frameProtocol")!r} '
            f'(expected {FRAME_PROTOCOL_ID})',
            code='FRAME_UNKNOWN_PROTOCOL',
        )
    if value.get('stream') not in ('request', 'response'):
        raise FrameError(
            f'frame: stream must be "request" or "response" (got {value.get("stream")!r})',
            code='FRAME_MALFORMED',
        )
    if value.get('encoding') != FRAME_ENCODING:
        # utf8-base64 is reserved in the schema but never emitted/accepted here.
        raise FrameError(
            f'frame: unsupported encoding {value.get("encoding")!r} '
            f'(only "utf8-slice" in 0.8.0)',
            code='FRAME_MALFORMED',
        )

    frame_id = value.get('id')
    if not _is_int(frame_id):
        raise FrameError(f'frame: id must be an integer (got {frame_id!r})', code='FRAME_MALFORMED')

    seq = value.get('seq')
    if not _is_int(seq) or seq < 0:
        raise FrameError(
            f'frame: seq must be a non-negative integer (got {seq!r})',
            code='FRAME_MALFORMED',
        )

    total = value.get('total')
    if not _is_int(total) or total < 1:
        raise FrameError(
            f'frame: total must be an integer >= 1 (got {total!r})',
            code='FRAME_MALFORMED',
        )

    total_bytes = value.get('totalBytes')
    if not _is_int(total_bytes) or total_bytes < 0:
        raise FrameError(
            f'frame: totalBytes must be a non-negative integer (got {total_bytes!r})',
            code='FRAME_MALFORMED',
        )

    data = value.get('data')
    if not isinstance(data, str):
        raise FrameError(
            f'frame: data must be a string (got {type(data).__name__})',
            code='FRAME_MALFORMED',
        )

    if seq >= total:
        raise FrameError(
            f'frame: seq {seq} out of range for total {total}',
            code='FRAME_MALFORMED',
        )

    return value


def _is_int(value: Any) -> bool:
    """True for a real integer (rejecting bool, which is an int subclass)."""
    return isinstance(value, int) and not isinstance(value, bool)


class Reassembler:
    """Accumulate ``tywrap-frame/1`` frames by ``id`` and reconstruct the string.

    Mirror of the TypeScript ``Reassembler`` class. A single instance handles
    many concurrent ids. Validation is enforced on every :meth:`accept`:

    - matching ``FRAME_PROTOCOL_ID`` on every frame;
    - consistent ``total`` / ``totalBytes`` across all frames of an id;
    - no duplicate ``seq``;
    - on completion, exactly ``total`` frames covering ``[0, total)``;
    - the concatenated payload's UTF-8 byte length equals ``totalBytes`` exactly;
    - the concatenated payload is valid UTF-8.

    Timed-out ids: the transport marks an id timed out via :meth:`discard`.
    Every subsequent frame for that id is dropped (returning ``None``) until its
    final frame arrives, at which point the id is forgotten so the slot can be
    reused. This prevents late multi-frame responses from desyncing the stream.
    """

    def __init__(self) -> None:
        # id -> {'total', 'totalBytes', 'slices': {seq: data}}
        self._streams: Dict[int, Dict[str, Any]] = {}
        self._discarded: set[int] = set()

    def accept(self, raw_frame: Any) -> Optional[str]:
        """Feed one frame.

        Returns the fully reassembled logical string when this frame completes
        the stream for its id, ``None`` if more frames are still needed (or the
        frame was dropped because its id is timed out).

        :raises FrameError: on any framing violation.
        """
        frame = parse_chunk_frame(raw_frame)
        frame_id = frame['id']
        seq = frame['seq']
        total = frame['total']
        total_bytes = frame['totalBytes']
        data = frame['data']

        # Late-frame discard: drop frames for a timed-out id; forget the id once
        # its declared final frame has been seen so the stream stays aligned and
        # the id can be reused.
        if frame_id in self._discarded:
            if seq == total - 1:
                self._discarded.discard(frame_id)
            return None

        state = self._streams.get(frame_id)
        if state is None:
            state = {'total': total, 'totalBytes': total_bytes, 'slices': {}}
            self._streams[frame_id] = state
        else:
            if state['total'] != total:
                del self._streams[frame_id]
                raise FrameError(
                    f'frame: total mismatch for id {frame_id} '
                    f'(saw {state["total"]}, frame says {total})',
                    code='FRAME_INCONSISTENT',
                )
            if state['totalBytes'] != total_bytes:
                del self._streams[frame_id]
                raise FrameError(
                    f'frame: totalBytes mismatch for id {frame_id} '
                    f'(saw {state["totalBytes"]}, frame says {total_bytes})',
                    code='FRAME_INCONSISTENT',
                )

        slices: Dict[int, str] = state['slices']
        if seq in slices:
            del self._streams[frame_id]
            raise FrameError(
                f'frame: duplicate seq {seq} for id {frame_id}',
                code='FRAME_DUPLICATE_SEQ',
            )
        slices[seq] = data

        if len(slices) < total:
            return None

        # All `total` frames present; the dense [0, total) range is guaranteed
        # because each seq is in range, unique, and there are exactly `total` of
        # them. Concatenate in seq order.
        del self._streams[frame_id]
        parts: List[str] = []
        for i in range(total):
            if i not in slices:
                # Unreachable given the count + uniqueness + range invariants,
                # but kept explicit rather than a silent gap.
                raise FrameError(
                    f'frame: missing seq {i} for id {frame_id}',
                    code='FRAME_SEQ_GAP',
                )
            parts.append(slices[i])
        payload = ''.join(parts)

        actual_bytes = utf8_byte_length(payload)
        if actual_bytes != total_bytes:
            raise FrameError(
                f'frame: reassembled byte length {actual_bytes} != declared '
                f'totalBytes {total_bytes} for id {frame_id}',
                code='FRAME_BYTES_MISMATCH',
            )

        # Strict UTF-8 validation. With utf8-slice the concatenation cannot
        # introduce invalid sequences (each slice is whole codepoints), but the
        # spec requires the check explicitly.
        try:
            payload.encode('utf-8').decode('utf-8')
        except UnicodeError as exc:  # pragma: no cover - defensive
            raise FrameError(
                f'frame: reassembled payload is not valid UTF-8 for id {frame_id}',
                code='FRAME_INVALID_UTF8',
            ) from exc

        return payload

    def discard(self, frame_id: int) -> None:
        """Mark an id as timed out / aborted.

        Drops any partial state immediately and discards every subsequent frame
        for this id until its declared final frame arrives. Idempotent.
        """
        self._streams.pop(frame_id, None)
        self._discarded.add(frame_id)

    def is_pending(self, frame_id: int) -> bool:
        """Whether any frame for ``frame_id`` is still being accumulated."""
        return frame_id in self._streams

    @property
    def pending_count(self) -> int:
        """Number of ids currently mid-reassembly (for diagnostics/tests)."""
        return len(self._streams)

File: src/runtime/frame-codec.ts
/**
 * Pure frame codec + reassembler for the `tywrap-frame/1` framing protocol.
 *
 * This module is the transport-agnostic core of large-payload chunking: it
 * fragments one complete logical JSON message into {@link ChunkFrame}s and
 * reassembles a stream of frames back into the original string. It performs NO
 * I/O and knows nothing about subprocesses, stdin/stdout, or timeouts — the
 * transport layer (W4/W5) wires these functions into the read/write loop.
 *
 * Encoding is `utf8-slice` (plan decision #6, docs/transport-framing.md): the
 * logical payload is already valid-UTF-8 JSON, so each frame's `data` is a raw
 * substring split on a UTF-8 codepoint boundary at or before `maxFrameBytes`
 * UTF-8 bytes. Reassembly is plain concatenation — no base64, no ~33% inflation,
 * no extra decode pass. A frame's `data` MUST NOT split a multi-byte UTF-8
 * sequence; {@link encodeFrames} guarantees this by snapping every boundary back
 * to the nearest codepoint boundary.
 *
 * The mirror implementation lives in `runtime/frame_codec.py`; the two MUST
 * agree byte-for-byte on the wire (see test/frame-codec.test.ts and
 * test/python/test_frame_codec.py for the cross-language parity vectors).
 *
 * @see docs/transport-framing.md
 */

import { BridgeProtocolError } from './errors.js';
import {
  FRAME_PROTOCOL_ID,
  type ChunkFrame,
  type ChunkFrameEncoding,
} from './transport.js';

// =============================================================================
// OPTIONS
// =============================================================================

/** Which logical stream a set of frames belongs to. */
export type FrameStream = ChunkFrame['stream'];

/** Options for {@link encodeFrames}. */
export interface EncodeFramesOptions {
  /** RPC correlation id, shared with the logical {@link ProtocolMessage.id}. */
  readonly id: number;

  /** Which logical stream these frames belong to (`request` or `response`). */
  readonly stream: FrameStream;

  /**
   * Maximum UTF-8 byte length of a single frame's `data` field. Each frame's
   * `data` is guaranteed to be at most this many UTF-8 bytes, snapped back to a
   * codepoint boundary so no multi-byte sequence is split. Must be a positive
   * integer large enough to hold at least one codepoint (>= 4 bytes).
   */
  readonly maxFrameBytes: number;
}

// =============================================================================
// UTF-8 HELPERS (pure, codepoint-boundary aware)
// =============================================================================

const utf8Encoder = new TextEncoder();
const utf8Decoder = new TextDecoder('utf-8', { fatal: true });

/** UTF-8 byte length of a single Unicode codepoint. */
function utf8ByteLengthOfCodePoint(codePoint: number): number {
  if (codePoint <= 0x7f) {
    return 1;
  }
  if (codePoint <= 0x7ff) {
    return 2;
  }
  if (codePoint <= 0xffff) {
    return 3;
  }
  return 4;
}

/** Exact UTF-8 byte length of a JS string (no allocation of the encoded bytes). */
export function utf8ByteLength(value: string): number {
  let bytes = 0;
  for (const ch of value) {
    bytes += utf8ByteLengthOfCodePoint(ch.codePointAt(0) ?? 0);
  }
  return bytes;
}

// =============================================================================
// ENCODE
// =============================================================================

/**
 * Fragment a complete logical JSON message into `tywrap-frame/1` frames.
 *
 * Splits `logicalJson` on UTF-8 codepoint boundaries so each frame's `data` is
 * at most `maxFrameBytes` UTF-8 bytes and never splits a multi-byte sequence.
 * `totalBytes` is the exact UTF-8 byte length of the full message; `total` is
 * the resulting frame count; `seq` is zero-based and dense.
 *
 * An empty `logicalJson` still produces exactly one (empty) frame so the
 * receiver always sees `total >= 1` and a well-formed stream.
 *
 * @throws BridgeProtocolError if `maxFrameBytes` is not a positive integer of at
 *   least 4 bytes (the worst-case single-codepoint size).
 */
export function encodeFrames(logicalJson: string, opts: EncodeFramesOptions): ChunkFrame[] {
  const { id, stream, maxFrameBytes } = opts;

  if (!Number.isInteger(maxFrameBytes) || maxFrameBytes < 4) {
    throw new BridgeProtocolError(
      `encodeFrames: maxFrameBytes must be an integer >= 4 (got ${String(maxFrameBytes)})`,
      { code: 'FRAME_BAD_MAX_BYTES' }
    );
  }
  // Validate the protocol fields locally too (parseChunkFrame enforces them on
  // the decode side, but catching misuse at the encoder fails fast and keeps the
  // two sides symmetric with the Python encoder).
  if (stream !== 'request' && stream !== 'response') {
    throw new BridgeProtocolError(
      `encodeFrames: stream must be 'request' or 'response' (got ${String(stream)})`,
      { code: 'FRAME_BAD_STREAM' }
    );
  }
  if (!Number.isInteger(id)) {
    throw new BridgeProtocolError(`encodeFrames: id must be an integer (got ${String(id)})`, {
      code: 'FRAME_BAD_ID',
    });
  }

  const totalBytes = utf8ByteLength(logicalJson);
  const encoding: ChunkFrameEncoding = 'utf8-slice';

  // Walk codepoints, accumulating UTF-8 bytes into the current slice until the
  // next codepoint would exceed maxFrameBytes; that boundary is, by
  // construction, a codepoint boundary so no multi-byte sequence is split.
  const slices: string[] = [];
  let current = '';
  let currentBytes = 0;
  for (const ch of logicalJson) {
    const chBytes = utf8ByteLengthOfCodePoint(ch.codePointAt(0) ?? 0);
    if (currentBytes + chBytes > maxFrameBytes && current.length > 0) {
      slices.push(current);
      current = '';
      currentBytes = 0;
    }
    current += ch;
    currentBytes += chBytes;
  }
  // Always emit a final slice (covers the empty-string case: one empty frame).
  slices.push(current);

  const total = slices.length;
  return slices.map((data, seq) => ({
    __tywrap_frame__: 'chunk',
    frameProtocol: FRAME_PROTOCOL_ID,
    stream,
    id,
    seq,
    total,
    totalBytes,
    encoding,
    data,
  }));
}

// =============================================================================
// REASSEMBLE
// =============================================================================

/** Per-id accumulation state inside the {@link Reassembler}. */
interface StreamState {
  total: number;
  totalBytes: number;
  /** Running UTF-8 byte count of slices received so far (memory bound). */
  bytesSoFar: number;
  /** seq -> frame data slice. */
  readonly slices: Map<number, string>;
}

/**
 * Validates that a raw value is a structurally well-formed `tywrap-frame/1` data
 * frame. Returns the frame typed, or throws a {@link BridgeProtocolError}.
 *
 * Only `__tywrap_frame__: 'chunk'` frames flow through reassembly; `'error'`
 * frames are a transport-layer concern handled above this module.
 */
export function parseChunkFrame(value: unknown): ChunkFrame {
  if (value === null || typeof value !== 'object') {
    throw new BridgeProtocolError('frame: expected an object', {
      code: 'FRAME_MALFORMED',
    });
  }
  const f = value as Record<string, unknown>;

  if (f.__tywrap_frame__ !== 'chunk') {
    throw new BridgeProtocolError(
      `frame: __tywrap_frame__ must be "chunk" (got ${JSON.stringify(f.__tywrap_frame__)})`,
      { code: 'FRAME_MALFORMED' }
    );
  }
  if (f.frameProtocol !== FRAME_PROTOCOL_ID) {
    throw new BridgeProtocolError(
      `frame: unknown frameProtocol ${JSON.stringify(f.frameProtocol)} (expected ${FRAME_PROTOCOL_ID})`,
      { code: 'FRAME_UNKNOWN_PROTOCOL' }
    );
  }
  if (f.stream !== 'request' && f.stream !== 'response') {
    throw new BridgeProtocolError(
      `frame: stream must be "request" or "response" (got ${JSON.stringify(f.stream)})`,
      { code: 'FRAME_MALFORMED' }
    );
  }
  if (f.encoding !== 'utf8-slice') {
    // utf8-base64 is reserved in the schema but never emitted/accepted in 0.8.0.
    throw new BridgeProtocolError(
      `frame: unsupported encoding ${JSON.stringify(f.encoding)} (only "utf8-slice" in 0.8.0)`,
      { code: 'FRAME_MALFORMED' }
    );
  }
  if (!Number.isInteger(f.id)) {
    throw new BridgeProtocolError(`frame: id must be an integer (got ${JSON.stringify(f.id)})`, {
      code: 'FRAME_MALFORMED',
    });
  }
  if (!Number.isInteger(f.seq) || (f.seq as number) < 0) {
    throw new BridgeProtocolError(
      `frame: seq must be a non-negative integer (got ${JSON.stringify(f.seq)})`,
      { code: 'FRAME_MALFORMED' }
    );
  }
  if (!Number.isInteger(f.total) || (f.total as number) < 1) {
    throw new BridgeProtocolError(
      `frame: total must be an integer >= 1 (got ${JSON.stringify(f.total)})`,
      { code: 'FRAME_MALFORMED' }
    );
  }
  if (!Number.isInteger(f.totalBytes) || (f.totalBytes as number) < 0) {
    throw new BridgeProtocolError(
      `frame: totalBytes must be a non-negative integer (got ${JSON.stringify(f.totalBytes)})`,
      { code: 'FRAME_MALFORMED' }
    );
  }
  if (typeof f.data !== 'string') {
    throw new BridgeProtocolError(`frame: data must be a string (got ${typeof f.data})`, {
      code: 'FRAME_MALFORMED',
    });
  }
  if ((f.seq as number) >= (f.total as number)) {
    throw new BridgeProtocolError(
      `frame: seq ${String(f.seq)} out of range for total ${String(f.total)}`,
      { code: 'FRAME_MALFORMED' }
    );
  }
  return f as unknown as ChunkFrame;
}

/**
 * Accumulates `tywrap-frame/1` frames by `id` and reconstructs the logical JSON
 * string once a stream is complete.
 *
 * A single instance handles many concurrent ids (each correlated by
 * {@link ChunkFrame.id}). Validation is enforced on every {@link accept} so a
 * malformed/duplicate/inconsistent frame fails fast rather than corrupting an
 * in-flight reassembly:
 *
 * - matching {@link FRAME_PROTOCOL_ID} on every frame;
 * - consistent `total` / `totalBytes` across all frames of an id;
 * - no duplicate `seq`;
 * - on completion, exactly `total` frames covering `[0, total)`;
 * - the concatenated payload's UTF-8 byte length equals `totalBytes` exactly;
 * - the concatenated payload decodes as strict UTF-8.
 *
 * Timed-out ids: the transport marks an id timed out via {@link discard}. Every
 * subsequent frame for that id is dropped (returning `null`) until its final
 * frame arrives, at which point the id is forgotten so the slot can be reused.
 * This prevents late multi-frame responses from desyncing the stream.
 */
/**
 * Defensive bounds on per-id reassembly state. The peer is the local Python
 * bridge, but a buggy or corrupt bridge must not grow memory without limit:
 * cap how many distinct ids may be mid-reassembly at once (fail loud past the
 * cap), and FIFO-bound the timed-out-id discard set so a long-lived process
 * that times out many chunked requests whose final frames never arrive (e.g.
 * the bridge died mid-stream) does not leak markers forever.
 */
const MAX_CONCURRENT_STREAMS = 1024;
const MAX_DISCARDED_IDS = 4096;

export class Reassembler {
  private readonly streams = new Map<number, StreamState>();
  private readonly discarded = new Set<number>();
  private readonly maxReassemblyBytes: number;
  private readonly expectedStream: FrameStream | undefined;

  /**
   * @param options.maxReassemblyBytes Reject (fail loud) any stream whose
   *   declared OR accumulated payload exceeds this many UTF-8 bytes, so a huge
   *   response cannot be buffered into memory before a higher layer's size cap
   *   would reject it anyway. Default: no limit (`Infinity`).
   * @param options.expectedStream If set, every frame must carry this `stream`
   *   direction; a frame for the other direction is rejected. The response and
   *   request reassemblers are distinct instances, so this is defense-in-depth.
   */
  constructor(options: { maxReassemblyBytes?: number; expectedStream?: FrameStream } = {}) {
    this.maxReassemblyBytes = options.maxReassemblyBytes ?? Number.POSITIVE_INFINITY;
    this.expectedStream = options.expectedStream;
  }

  /**
   * Feed one frame. Returns the fully reassembled logical string when this
   * frame completes the stream for its id, `null` if more frames are still
   * needed (or the frame was dropped because its id is timed out).
   *
   * @throws BridgeProtocolError on any framing violation (malformed frame,
   *   wrong stream direction, declared/accumulated payload over the cap,
   *   duplicate `seq`, inconsistent `total`/`totalBytes`, byte-count mismatch,
   *   invalid UTF-8, or unknown `frameProtocol`).
   */
  accept(rawFrame: unknown): string | null {
    const frame = parseChunkFrame(rawFrame);
    const { id, seq, total, totalBytes, data, stream } = frame;

    if (this.expectedStream !== undefined && stream !== this.expectedStream) {
      throw new BridgeProtocolError(
        `frame: unexpected stream '${stream}' for id ${id} (this reassembler handles '${this.expectedStream}')`,
        { code: 'FRAME_WRONG_STREAM' }
      );
    }

    // Late-frame discard: drop frames for a timed-out id; forget the id once its
    // declared final frame has been seen so the stream stays aligned and the id
    // can be reused.
    if (this.discarded.has(id)) {
      if (seq === total - 1) {
        this.discarded.delete(id);
      }
      return null;
    }

    let state = this.streams.get(id);
    if (state === undefined) {
      if (this.streams.size >= MAX_CONCURRENT_STREAMS) {
        throw new BridgeProtocolError(
          `frame: too many concurrent reassembly streams (>= ${MAX_CONCURRENT_STREAMS}); refusing to buffer id ${id}`,
          { code: 'FRAME_TOO_MANY_STREAMS' }
        );
      }
      // Reject early: a declared payload past the cap can never be accepted by
      // the higher-layer size guard, so refuse to buffer it instead of OOMing.
      if (totalBytes > this.maxReassemblyBytes) {
        throw new BridgeProtocolError(
          `frame: declared payload ${totalBytes} bytes exceeds max reassembly ${this.maxReassemblyBytes} bytes for id ${id}`,
          { code: 'FRAME_PAYLOAD_TOO_LARGE' }
        );
      }
      state = { total, totalBytes, bytesSoFar: 0, slices: new Map<number, string>() };
      this.streams.set(id, state);
    } else {
      if (state.total !== total) {
        this.streams.delete(id);
        throw new BridgeProtocolError(
          `frame: total mismatch for id ${id} (saw ${state.total}, frame says ${total})`,
          { code: 'FRAME_INCONSISTENT' }
        );
      }
      if (state.totalBytes !== totalBytes) {
        this.streams.delete(id);
        throw new BridgeProtocolError(
          `frame: totalBytes mismatch for id ${id} (saw ${state.totalBytes}, frame says ${totalBytes})`,
          { code: 'FRAME_INCONSISTENT' }
        );
      }
    }

    if (state.slices.has(seq)) {
      this.streams.delete(id);
      throw new BridgeProtocolError(`frame: duplicate seq ${seq} for id ${id}`, {
        code: 'FRAME_DUPLICATE_SEQ',
      });
    }
    state.slices.set(seq, data);

    // Running memory bound: a peer that under-declares totalBytes then overshoots
    // is caught here (and the exact-count check at completion catches the rest)
    // before the full payload is buffered.
    state.bytesSoFar += utf8ByteLength(data);
    if (state.bytesSoFar > this.maxReassemblyBytes) {
      this.streams.delete(id);
      throw new BridgeProtocolError(
        `frame: accumulated payload exceeds max reassembly ${this.maxReassemblyBytes} bytes for id ${id}`,
        { code: 'FRAME_PAYLOAD_TOO_LARGE' }
      );
    }

    if (state.slices.size < total) {
      return null;
    }

    // All `total` frames present; the dense [0, total) range is guaranteed
    // because each seq is in range, unique, and there are exactly `total` of
    // them. Concatenate in seq order.
    this.streams.delete(id);
    let payload = '';
    for (let i = 0; i < total; i += 1) {
      const slice = state.slices.get(i);
      if (slice === undefined) {
        // Unreachable given the count + uniqueness + range invariants above,
        // but kept explicit rather than a silent gap.
        throw new BridgeProtocolError(`frame: missing seq ${i} for id ${id}`, {
          code: 'FRAME_SEQ_GAP',
        });
      }
      payload += slice;
    }

    const actualBytes = utf8ByteLength(payload);
    if (actualBytes !== totalBytes) {
      throw new BridgeProtocolError(
        `frame: reassembled byte length ${actualBytes} != declared totalBytes ${totalBytes} for id ${id}`,
        { code: 'FRAME_BYTES_MISMATCH' }
      );
    }

    // Strict UTF-8 validation: re-encode then decode in fatal mode. With
    // utf8-slice the concatenation cannot introduce invalid sequences (each
    // slice is whole codepoints), but the spec requires the check explicitly.
    try {
      utf8Decoder.decode(utf8Encoder.encode(payload));
    } catch (cause) {
      throw new BridgeProtocolError(`frame: reassembled payload is not valid UTF-8 for id ${id}`, {
        code: 'FRAME_INVALID_UTF8',
        cause,
      });
    }

    return payload;
  }

  /**
   * Mark an id as timed out / aborted. Any partial state is dropped immediately
   * and every subsequent frame for this id is discarded until its declared
   * final frame arrives. Idempotent.
   */
  discard(id: number): void {
    this.streams.delete(id);
    this.discarded.add(id);
    // Bound the discard set (FIFO): a timed-out id whose declared final frame
    // never arrives would otherwise linger forever. Set iteration order is
    // insertion order, so the first key is the oldest marker.
    if (this.discarded.size > MAX_DISCARDED_IDS) {
      const oldest = this.discarded.values().next().value;
      if (oldest !== undefined) {
        this.discarded.delete(oldest);
      }
    }
  }

  /** Whether any frame for `id` is still being accumulated (not yet complete). */
  isPending(id: number): boolean {
    return this.streams.has(id);
  }

  /** Number of ids currently mid-reassembly (for diagnostics/tests). */
  get pendingCount(): number {
    return this.streams.size;
  }

  /**
   * Number of timed-out ids whose late frames are still being dropped
   * (FIFO-bounded by `MAX_DISCARDED_IDS`; for diagnostics/tests).
   */
  get discardedCount(): number {
    return this.discarded.size;
  }
}
